src/permutation.cpp

```diff
@@ -0,0 +1,248 @@
+#include "permutation.h"
+
+#include <algorithm>
+#include <sstream>
+#include <stdexcept>
+#include <vector>
+
+namespace Simhash {
+
+    std::vector<std::vector<hash_t> > Permutation::choose(
+            const std::vector<hash_t>& population, size_t r)
+    {
+        // This algorithm is cribbed from python's itertools page.
+        size_t n = population.size();
+        if (r > n)
+        {
+            throw std::invalid_argument("R cannot be greater than population size.");
+        }
+
+        std::vector<size_t> indices(r);
+        for (size_t i = 0; i < r; ++i)
+        {
+            indices[i] = i;
+        }
+
+        std::vector<std::vector<hash_t> > results;
+        std::vector<hash_t> result(r);
+
+        for (size_t i = 0; i < r; ++i)
+        {
+            result[i] = population[indices[i]];
+        }
+        results.push_back(result);
+
+        while (true)
+        {
+            int i = r - 1;
+            for (; i >= 0; --i)
+            {
+                if (indices[i] != i + n - r)
+                {
+                    break;
+                }
+            }
+            if (i < 0)
+            {
+                return results;
+            }
+
+            indices[i] += 1;
+            for (size_t j = i + 1; j < r; ++j)
+            {
+                indices[j] = indices[j-1] + 1;
+            }
+            for (size_t j = 0; j < r; ++j)
+            {
+                result[j] = population[indices[j]];
+            }
+            results.push_back(result);
```

Comment:
Actually, this is `O(population.size() C r)`, and I'd rather not implement `nChooseK` -- is there an implementation readily available?